python/hail/dataset.py

```diff
@@ -915,6 +915,278 @@ def annotate_variants_vds(self, other, expr=None, root=None):
+                            let canonical_transcripts = va.vep.transcript_consequences.filter(t => t.canonical == 1) in
+                            if (canonical_transcripts.exists(t => t.consequence_terms.toSet.contains(va.gene.consequence)))
+                                canonical_transcripts.find(t => t.consequence_terms.toSet.contains(va.gene.consequence)).gene_symbol
+                            else
+                                va.vep.transcript_consequences.find(t => t.consequence_terms.toSet.contains(va.gene.consequence)).gene_symbol,
+                        va.gene.gene_id =
+                            let canonical_transcripts = va.vep.transcript_consequences.filter(t => t.canonical == 1) in
+                            if (canonical_transcripts.exists(t => t.consequence_terms.toSet.contains(va.gene.consequence)))
+                                canonical_transcripts.find(t => t.consequence_terms.toSet.contains(va.gene.consequence)).gene_id
+                            else
+                                va.vep.transcript_consequences.find(t => t.consequence_terms.toSet.contains(va.gene.consequence)).gene_symbol,
+                        va.transcript_id = let canonical_transcripts = va.vep.transcript_consequences.filter(t => t.canonical == 1) in
+                            if (canonical_transcripts.exists(t => t.consequence_terms.toSet.contains(va.gene.consequence)))
+                                canonical_transcripts.find(t => t.consequence_terms.toSet.contains(va.gene.consequence)).transcript_id
+                            else
+                                va.vep.transcript_consequences.find(t => t.consequence_terms.toSet.contains(va.gene.consequence)).transcript_id
+                        """
+                    )
+                )
+
+            # drop VEP annotations if not specifically requested
+            if not veps:
+                self = self.annotate_variants_expr('va = drop(va, vep)')
+
+            # subset VEP annotations if needed
+            subset = ','.join([x for x in veps if x != 'va.vep' and x != 'va'])
+            if subset:
+                self = self.annotate_variants_expr('va.vep = select(va.vep, {})'.format(subset))
+
+        # iterate through files, selected annotations from each file
+        for file, expr in file_exprs.iteritems():
+
+            # skip VEP annotations
+            if file.startswith('gs://annotationdb/vep/'):
+                continue
+
+            # if database file is a VDS
+            if file.endswith('.vds'):
+
+                # annotate analysis VDS with database VDS
+                self = self.annotate_variants_vds(self.hc.read('{}'.format(file)), expr = expr)
+
+            # if database file is a keytable
+            elif file.endswith('.kt'):
+
+                # join on gene symbol for gene annotations
+                if file.startswith('gs://annotationdb/gene/'):
+                    if gene_key:
```

Comment:
The gene keytables in the database are keyed by a single gene symbol string. The `gene_key` parameter here would be the path to a single annotation already in the VDS. `gene_key = 'va.my_gene'` or something, where `va.my_gene` is a string.